src/libawkward/kernel.cpp

```diff
@@ -13,12 +13,1316 @@
+
+    typedef Error (func_awkward_cuda_ptrU8_dealloc_t)(const uint8_t *ptr);
+    func_awkward_cuda_ptrU8_dealloc_t *func_awkward_cuda_ptrU8_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptrU8_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptrU8_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptrU8_dealloc)(p));
+#endif
+  }
+
+  template<>
+  void cuda_array_deleter<int16_t>::operator()(int16_t const *p) {
+#ifndef _MSC_VER
+    auto handle = acquire_handle(kernel::Lib::cuda_kernels);
+
+    typedef Error (func_awkward_cuda_ptr16_dealloc_t)(const int16_t *ptr);
+    func_awkward_cuda_ptr16_dealloc_t *func_awkward_cuda_ptr16_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptr16_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptr16_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptr16_dealloc)(p));
+#endif
+  }
+
+  template<>
+  void cuda_array_deleter<uint16_t>::operator()(uint16_t const *p) {
+#ifndef _MSC_VER
+    auto handle = acquire_handle(kernel::Lib::cuda_kernels);
+
+    typedef Error (func_awkward_cuda_ptrU16_dealloc_t)(const uint16_t *ptr);
+    func_awkward_cuda_ptrU16_dealloc_t *func_awkward_cuda_ptrU16_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptrU16_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptrU16_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptrU16_dealloc)(p));
+#endif
+  }
+
+  template<>
+  void cuda_array_deleter<int32_t>::operator()(int32_t const *p) {
+#ifndef _MSC_VER
+    auto handle = acquire_handle(kernel::Lib::cuda_kernels);
+
+    typedef Error (func_awkward_cuda_ptr32_dealloc_t)(const int32_t *ptr);
+    func_awkward_cuda_ptr32_dealloc_t *func_awkward_cuda_ptr32_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptr32_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptr32_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptr32_dealloc)(p));
+#endif
+  }
+
+  template<>
+  void cuda_array_deleter<uint32_t>::operator()(uint32_t const *p) {
+#ifndef _MSC_VER
+    auto handle = acquire_handle(kernel::Lib::cuda_kernels);
+
+    typedef Error (func_awkward_cuda_ptrU32_dealloc_t)(const uint32_t *ptr);
+    func_awkward_cuda_ptrU32_dealloc_t *func_awkward_cuda_ptrU32_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptrU32_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptrU32_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptrU32_dealloc)(p));
+#endif
+  }
+
+  template<>
+  void cuda_array_deleter<int64_t>::operator()(int64_t const *p) {
+#ifndef _MSC_VER
+    auto handle = acquire_handle(kernel::Lib::cuda_kernels);
+
+    typedef Error (func_awkward_cuda_ptr64_dealloc_t)(const int64_t *ptr);
+    func_awkward_cuda_ptr64_dealloc_t *func_awkward_cuda_ptr64_dealloc =
+      reinterpret_cast<func_awkward_cuda_ptr64_dealloc_t *>
+      (acquire_symbol(handle, "awkward_cuda_ptr64_dealloc"));
+
+    util::handle_error((*func_awkward_cuda_ptr64_dealloc)(p));
+#endif
```

Comment:
As I said earlier, all of these identical snippets of code need to be gathered into a single utility function. Maintaining so many identical code blocks would be a nightmare!

Reply:
Do you have anything in mind as to how I can reduce this duplication? For starts, the `symbol_lookups` like "awkward_cuda_ptr64_dealloc" are very specific. So although, I could use templates and use string manipulation to get the symbol, I think it would affect the readability of the code base. I have already, replaced the `dlopen` calls everywhere with one single utility function and also `dlsym`. It handles all the string checking and exception handling.

Reply:
Oh, I see—it's already reduced down from what it could be. This will take some thought. You need to keep the choice of function signature (templated type) in sync with its lookup name (string).

There are several things here that are dangerous for maintainability:

   * `#ifndef`: preprocessor macros are brittle: the C++ compiler can't reason about them and editors usually can't either. Having one or two of them around is okay, if necessary, but having hundreds of them opens the door to confusing errors if an `#endif` gets misplaced. I'm a little confused about how this compiles properly (for cpu-kernels) on Windows, since there's no `#else` in these conditionals. Anyway, it should be possible to wrap them up in a function by creating our own `handle` type (which is `auto` here, so I can't see what it is) that defers to something trivial in the Windows case (with a single `#ifndef` conditional).
   * `reinterpret_cast` is dangerous, again because the C++ compiler can't tell us if it's right. I originally introduced the templatized kernels in `util::` to remove `reinterpret_casts` from the calling sites. (That was the motivation, more than convenience.) Having hundreds of `reinterpret_casts` that have to line up with the correct `func_awkward_cuda_ptr64_dealloc_t` types and `"awkward_cuda_ptr64_dealloc"` strings—it's just dangerous.

Maybe the right way to fix this is with a preprocessor macro ("fight fire with fire"). The compile-time typedef and the run-time string are very different things, but they have the same string in the text of the program. Perhaps these could all be replaced by

```c++
RUN_KERNEL(func_awkward_cuda_ptr64_dealloc)
```

with some preprocessor operations in `#define RUN_KERNEL(function_name)` that construct the compile-time typedef and the run-time string from the `function_name` argument. That wouldn't remove the `#ifdef` for Windows or the `reinterpret_cast` (because `dlopen` is a C interface without much type-safety), but it would isolate them to the `#define RUN_KERNEL` definition, a single location in the code where we can make sure these dangerous things are used correctly, once and for all. Having hundreds of macro-calls to `RUN_KERNEL` isn't bad.

Presumably, this could be done with C++ metaprogramming, but I've found C++'s metaprogramming language to be very hard to understand. (Metaprogramming in other languages, like Scala and Python, is much easier: it's C++'s language choices that are the problems.) C preprocessor macros lack the safety guarantees of C++ metaprogramming, but understanding it is better than automated safety, particularly since the unsafe `reinterpret_cast` probably has to be there because `dlopen` isn't type-safe.

Reply:
I am not a whole lot familiar with extreme use-cases of macros, and maybe I am doing it wrong but I can't nest #define and #ifndef, that's a first. After that, since `dlopen` returns a opaque pointer, using it essentially requires to cast it before using it, and I doubt there's any workaround that doesn't use `reinterpret_cast`. Also, the parameters(function arguments to the kernel) in the typedef are different, but from a little search I think using `_VA_ARGS_` could work. 
It would be very helpful for me, if you could provide me with a small example as to how the macro is going to be defined.

Reply:
I know that you can put a `#define` inside of an `#ifndef`, though it might not be possible to do it the other way around.

Macros are essentially string-processing on the text of the source file (which is why they provide so few guarantees), so if something depends on a function that could be derived from some code, you can pass in the name of the function and use it both to select the function's arguments and use it as the string that gets passed to `dlopen`. I can't give an example because I haven't thought through the solution (and I also don't know the preprocessor instructions for concatenating strings to build the code you want), but you may be able to apply a little creativity here. The macro doesn't necessary have to call the function: it might just construct the function and the actual argument passing might happen on the output of the macro (making var-args unnecessary).

The point of this is to be simple, so if trying to write this is getting out of hand, then it's better not to. Basically, you want to find any way to reduce the code duplication (a.k.a. "boilerplate") so that what we see in each function is pretty much just passing through the function arguments by name. If there has to be boilerplate, making it regular enough to edit by "replace all" could be acceptable, because under certain conditions that's not prone to error. But anything that has intricate invariants to maintain is dangerous.

I'll think about it, too. This can't be the first time someone has needed to glue a large codebase accessible through `dlopen`.

------------------

Is there a way to ask for "`type of <FUNCTION-NAME-HERE>`"? Maybe then the so-called "`cpu-kernels/*.h`" could actually become generic "`kernels/*.h`" with the cpu-kernel implementations included to satisfy the loader (they already are) and you could get the full type definition with some "`type of ...`" on the "`kernels/*.h`" version, but then cast that on the function that you get from `dlopen`, which would actually be the CUDA version.

After all, the cuda-kernels are supposed to satisfy exactly the same interface as the cpu-kernels, so there only needs to be one set of header files, which are in scope in the `kernel::` namespace.

Reply:
Yeah, I found some use cases of `dlopen` and `dlsym` in the Pytorch and Tensorflow Libraries. I am currently studying them.

On a completely different note, if isolation the `_MSC_VER` flag to one place is the real aim here(I know generalizing the entire thing is too, but the problem is that not all kernels return an ERROR(some return pointers) and that makes macros a bit difficult), I can isolate them in two places 1. `acquire_handle` and 2. `acquire_symbol`. I can skip putting those flags everywhere since it becomes redundant(I should have done this earlier).

Reply:
Oh, for the second part of your comment, I can use `decltype(awkward_FnName)` to deduce the type.

Reply:
> if isolation the _MSC_VER flag to one place is the real aim here

That's part of it, though the overall goal is to reduce all brittle duplication. Hiding them in macros that `BEGIN` and `END` a section isn't terrible.

> the problem is that not all kernels return an ERROR(some return pointers) and that makes macros a bit difficult)

Don't worry about that! Write the macros for the majority of kernels (which do return `Error`) and treat the others as special cases. The point is to make most of it cleaner, and if you can do that by macroizing only the most regular, that's fine. (There are hundreds of kernels that return `Error` and only a handful that return pointers or nothing.)

For the last part, `decltype` looks like exactly what you want.

Reply:
So leaving, `copy_to` and `get_device_*` functions, I was able to make all the kernels use the new macro. I had to introduce a few new `cpu-kernels`(which was currently being done directly on the C++ side) but that adds a bit more symmetry to the entire thing.